libs/providers/flagd/src/lib/service/grpc/service.ts

```diff
@@ -2,111 +2,246 @@ import * as grpc from '@grpc/grpc-js';
 import { ServiceClient } from '../../../proto/ts/schema/v1/schema.client';
 import { Config } from '../../configuration';
+import {
+  BASE_EVENT_STREAM_RETRY_BACKOFF_MS,
+  DEFAULT_MAX_CACHE_SIZE,
+  DEFAULT_MAX_EVENT_STREAM_RETRIES,
+  EVENT_CONFIGURATION_CHANGE,
+  EVENT_PROVIDER_READY,
+} from '../../constants';
+import { FlagdProvider } from '../../flagd-provider';
 import { Service } from '../service';
 
+type AnyResponse =
+  | ResolveBooleanResponse
+  | ResolveStringResponse
+  | ResolveIntResponse
+  | ResolveFloatResponse
+  | ResolveObjectResponse;
+type AnyRequest =
+  | ResolveBooleanRequest
+  | ResolveStringRequest
+  | ResolveIntRequest
+  | ResolveFloatRequest
+  | ResolveObjectRequest;
+
+interface FlagChange {
+  type: 'delete' | 'write' | 'update';
+  source: string;
+  flagKey: string;
+}
+
+export interface FlagChangeMessage {
+  flags?: { [key: string]: FlagChange };
+}
+
 // see: https://grpc.github.io/grpc/core/md_doc_statuscodes.html
 export const Codes = {
   InvalidArgument: 'INVALID_ARGUMENT',
   NotFound: 'NOT_FOUND',
   DataLoss: 'DATA_LOSS',
-  Unavailable: 'UNAVAILABLE'
+  Unavailable: 'UNAVAILABLE',
 } as const;
 
 export class GRPCService implements Service {
-  client: ServiceClient;
+  private _client: ServiceClient;
+  private _cache: LRU<string, ResolutionDetails<FlagValue>> | undefined;
+  private _cacheEnabled = false;
+  private _streamAlive = false;
+  private _streamConnectAttempt = 0;
+  private _streamConnectBackoff = BASE_EVENT_STREAM_RETRY_BACKOFF_MS;
+  private _maxEventStreamRetries;
+  private get _cacheActive() {
+    // the cache is "active" (able to be used) if the config enabled it, AND the gRPC stream is live
+    return this._cacheEnabled && this._streamAlive;
+  }
+
+  // default to false here - reassigned in the constructor if we actaully need to connect
+  readonly streamConnection = Promise.resolve(false);
 
-  constructor(config: Config, client?: ServiceClient) {
+  constructor(config: Config, client?: ServiceClient, private logger?: Logger) {
     const { host, port, tls, socketPath } = config;
-    this.client = client
+    this._maxEventStreamRetries = config.maxEventStreamRetries || DEFAULT_MAX_EVENT_STREAM_RETRIES;
+    this._client = client
       ? client
       : new ServiceClient(
           new GrpcTransport({
             host: socketPath ? `unix://${socketPath}` : `${host}:${port}`,
-            channelCredentials: tls
-              ? grpc.credentials.createSsl()
-              : grpc.credentials.createInsecure(),
+            channelCredentials: tls ? grpc.credentials.createSsl() : grpc.credentials.createInsecure(),
           })
         );
+
+    // for now, we only need streaming if the cache is enabled (will need to be pulled out once we support events)
+    if (config.cache === 'lru') {
+      this._cacheEnabled = true;
+      this._cache = new LRU({ maxSize: config.maxCacheSize || DEFAULT_MAX_CACHE_SIZE, sizeCalculation: () => 1 });
```

Comment:
It looks like if a user sets the `maxCacheSize` to zero it will use the `DEFAULT_MAX_CACHE_SIZE`. That might be slightly confusing. Perhaps we should disable the cache in the situation. What do you think?

Reply:
I think we have a clear way to disable cache with `'disabled'`.

Reply:
Yeah, that's fair. I'm fine with this then.